### PR TITLE
feat(QEff: Attn): add KV & Q blocking strategies for causal LMs

### PR DESCRIPTION
**Pull request by vbaddi**
**Summary**

- This PR introduces QEfficient attention blocking for causal language models, adding KV and Q blocking strategies that preserve numerical outputs and scalability at long sequence lengths.

**Key Features**

- KV Blocking: Blocked compute over KV cache along sequence dim.
- Q Blocking: Blocked compute over query sequence.
- Currently, Configurable via qaic_config (explicit or auto).
- Moved transforms for KV blocking, Q blocking and Head blocking to `.transform()` method called at compile time, since these transforms are dependent on compilation parameters like CL. Other transforms still in `.from_pretrained()` for now. 

**WIP**
- [X] Implement the Auto Blocking
- [ ] Move the transforms structure from `.from_pretrained()` to `.compile()` and keep a depreciation warning at `.from_pretrained()` level. ; **Not targeted for this PR**
- [X] Unit tests
- 
++ @kdulla

### CODE REVIEW

File: QEfficient/base/modeling_qeff.py [hunk 1]
```diff
@@ -366,6 +462,9 @@ def _compile(
         offload_pt_weights: Optional[bool] = True,
         enable_chunking: Optional[bool] = False,
         retain_full_kv: Optional[bool] = None,
+        disable_blocking: Optional[bool] = True,
+        blocking_mode: Optional[str] = "hqkv",
+        vtcm_ratio: Optional[float] = 0.75,
```
Comment (ochougul) [inline QEfficient/base/modeling_qeff.py]: Is it possible to take `qaic_config` as input here as well.
If user passes it, we override the old self.model.qaic_config else we let it be?

**Review by ochougul — CHANGES_REQUESTED** [review]
review WIP

File: QEfficient/base/modeling_qeff.py [hunk 2]
```diff
@@ -348,6 +357,93 @@ def get_onnx_path(
...
+        seq_len: Optional[int] = None,
+        bs: Optional[int] = 1,
+        num_devices: int = 1,
+        disable_blocking: Optional[bool] = True,
+        blocking_mode: Optional[str] = "hqkv",
+        vtcm_ratio: Optional[float] = 0.75,
+        **compiler_options,
+    ):
+        # Apply the transformations that are dependent on compilation parameters
+        if getattr(self.model, "qaic_config", None) is None and not disable_blocking:
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py]: lets move the blocking config creation logic out of transform method, you can keep probably keep the logic in blocking_configurator.py

**Review by quic-rishinr — CHANGES_REQUESTED** [review]

**Review by ochougul — COMMENTED** [review]
please push an example file

File: examples/text_generation/blocked_attention_inference.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    tokenizer = AutoTokenizer.from_pretrained(args.model_name)
+    model = QEFFAutoModelForCausalLM.from_pretrained(args.model_name, num_hidden_layers=2)
+
+    # setup qaic config to enable blocking, ensure 4 or more device ids are passed
+    qaic_config = {"enable_blocking": True, "blocking_mode": "kv"}
+
+    # Compile the model
+    qpc_path = model.compile(
+        prefill_seq_len=args.prefill_seq_len,
+        ctx_len=args.ctx_len,
```
Comment (vaibverm) [inline examples/text_generation/blocked_attention_inference.py]: When the transform for KV Blocking moves from the .from_pretrained() to .compile(), is the intent to export and then compile the model with full CL value (32K in this example)?

**Review by anujgupt-github — CHANGES_REQUESTED** [review]

**Comment by vbaddi** [comment]
duplicate #908
